tests/python/pants_test/engine/test_isolated_process.py

```diff
@@ -84,22 +154,125 @@ def java_sources_to_javac_args(sources_snapshot, out_dir):
+
+
+@rule(ExecuteProcessRequest, [Select(JavacVersionCommand)])
+def process_request_from_javac_version(javac_version_command):
+  yield ExecuteProcessRequest(
+    argv=javac_version_command.gen_argv(),
+    env=[],
+    input_files_digest=EMPTY_FINGERPRINT,
     digest_length=0)
 
 
-def process_request_java_args_from_java_sources(binary, sources_snapshot, out_dir):
-  env = []
-  return ExecuteProcessRequest(
-    argv=binary.prefix_of_command() + tuple(f.path for f in sources_snapshot.files),
-    input_files_digest=str(sources_snapshot.fingerprint),
-    digest_length=sources_snapshot.digest_length,
-    env=env)
+class JavacVersionOutput(datatype('JavacVersionOutput', [
+    'exit_code',
+    'version_output',
+])): pass
+
+
+@rule(JavacVersionOutput, [Select(JavacVersionCommand)])
+def get_javac_version_output(javac_version_command):
+  javac_version_proc_req = yield Get(
+    ExecuteProcessRequest, JavacVersionCommand, javac_version_command)
+  javac_version_proc_result = yield Get(
+    ExecuteProcessResult, ExecuteProcessRequest, javac_version_proc_req)
+  yield JavacVersionOutput(
+    exit_code=javac_version_proc_result.exit_code,
+    version_output=javac_version_proc_result.stderr,
+  )
+
+
+class JavacCompileCommand(Javac):
+
+  def gen_argv(self, snapshots):
+    snapshot_file_paths = []
+    for s in snapshots:
+      snapshot_file_paths.extend(f.path for f in s.files)
+
+    return (self.bin_path,) + tuple(snapshot_file_paths)
+
+
+class JavacSources(datatype('JavacSources', ['globs'])):
+
+  def __new__(cls, globs):
+
+    if not isinstance(globs, PathGlobs):
+      raise ValueError('globs should be an instance of PathGlobs')
+
+    return super(JavacSources, cls).__new__(cls, globs)
+
+
+@rule(PathGlobs, [Select(JavacSources)])
+def javac_sources_to_globs(javac_sources):
+  yield javac_sources.globs
+
+
+class JavacCompileRequest(datatype('JavacCompileRequest', [
+    'javac_compile_command',
+    'javac_sources',
+])):
+
+  def __new__(cls, javac_compile_command, javac_sources):
+
+    if not isinstance(javac_compile_command, JavacCompileCommand):
+      raise ValueError(
+        'javac_compile_command should be an instance of JavacCompileCommand')
+    if not isinstance(javac_sources, JavacSources):
+      raise ValueError(
+        'javac_sources should be an instance of JavacSources')
+
+    return super(JavacCompileRequest, cls).__new__(
+      cls, javac_compile_command, javac_sources)
+
+
+@rule(ExecuteProcessRequest, [Select(JavacCompileRequest)])
+def javac_compile_sources_execute_process_request(javac_compile_req):
+  javac_compiler = javac_compile_req.javac_compile_command
+  sources_snapshot = yield Get(
+    Snapshot, JavacSources, javac_compile_req.javac_sources)
+  yield ExecuteProcessRequest.create_from_snapshot(
+    argv=javac_compiler.gen_argv([sources_snapshot]),
+    env=tuple(),
+    snapshot=sources_snapshot,
+  )
+
+
+class JavacCompileResult(datatype('JavacCompileResult', [
+    'exit_code',
```

Comment:
Again, exit_code seems strange here...

Reply:
Agreed, left this as an empty datatype for now until we can use it to wrap process output file snapshots.

Reply:
...and just added a TODO to record that.